platform/sensor-stream-manager/sensor-stream-manager.js

```diff
@@ -1,10 +1,111 @@
-//TODO: add mqtt-data-collector logic to SSM.
+const fs = require("fs-extra");
+const mqtt = require("mqtt");
+const utils = require("../../utils");
+
+// TODO: add mqtt-data-collector logic to SSM.
 const MessagingService = require('../messaging-service');
 
 const serviceName = process.env.SERVICE_NAME;
 const messagingService = new MessagingService(serviceName);
 
+// get gateway's ip
+const gatewayIp = utils.getIPAddress();
+if(!gatewayIp) {
+  console.error("No IP address found. Please ensure the config files are set properly.");
+  process.exit(1);
+}
+console.log(`gateway's ip is ${gatewayIp}`);
+
+
+// sensorStream stores the sensor id and application topic mapping
+// the key is sensor id and the value is an array of application's topics
+// SSM will use sensorStream to push data
+// check the example below
+// sensorStream = {
+//     "sensor1-id": [ '5ec23c4a30802a720dfea97d' ],
+//     "sensor2-id": [ '5ec23c4a30802a720dfea97d' ]
+// }
+const sensorStream = {};
+
+// TODO: policy
+const policy = {};
+
+// connect to localhost mqtt broker
+const mqttClient = mqtt.connect("mqtt://localhost");
+mqttClient.on('connect', () => {
+    // subscribe to localhost "gateway-data" topic
+    mqttClient.subscribe("gateway-data", (err) => {
+        if (err) {
+            console.error(`mqtt client subscribe ${mqttTopic} failed.`);
+            console.error(err);
+        } else {
+            console.log(`subscribe to "gateway-data" topic successfully!`);
+        }
+    });
+});
+
+// when sensor stream data is published to "gateway-data" topic
+// SSM will check policy and publish the data to application's topic
+mqttClient.on('message', (topic, message) => {
+    const payload = JSON.parse(message.toString());
+    const sensorId = payload["_meta"]["device_id"];
+    if(sensorId in sensorStream) {
+        for(let topic of sensorStream[sensorId]) {
+            // TODO: check policy
+
+            // publish to application's topic
+            mqttClient.publish(topic, JSON.stringify(payload));
+        }
+    }
+});
+
 messagingService.listenForEvent('connect-to-socket', (message) => {
     const payload = message.data;
     const wsAddress = payload["ws-address"];
+});
+
+// SSM receives application's process instance and metadata from app-manager
+// this listener will store the application's topic and sensor stream requirement
+messagingService.listenForEvent('app-deployment', message => {
+    // appData = {
+    //     "app": {
+    //         "app": process-instance,
+    //         "topic": "app-topic",
+    //         "path": "app-path"
+    //     },
+    //     "metadataPath": "metadata-path"
+    // };
+    let appData = message.data;
+    if(appData["app"] && appData["metadataPath"]) {
+        // load application's metadata
+        let metadata = fs.readJsonSync(appData["metadataPath"]);
```

Comment:
why do you need readJsonSync?

Reply:
I need to read metadata to check the application's sensor requirement.